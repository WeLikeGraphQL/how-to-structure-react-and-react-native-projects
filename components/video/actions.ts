export const actionTypes = {
  PLAY_VIDEO: "@GRAPHAN/VIDEO_PLAY",
  STOP_VIDEO: "@GRAPHAN/VIDEO_STOP"
};

export const actionCreators = {
  playVideoAction: payload => ({
    type: actionTypes.PLAY_VIDEO,
    payload
  }),
  stopVideoAction: payload => ({
    type: actionTypes.STOP_VIDEO,
    payload
  })
};
