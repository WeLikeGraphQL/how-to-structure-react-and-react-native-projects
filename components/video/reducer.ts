import { actionTypes } from './actions'

const initialState = {}

const stopAllVideos = (state = {}) => {
  return Object.keys(state).reduce((acc, key) => {
    acc[key] = 'stopped'
    return acc
  }, {})
}

function reducer(state = initialState, action) {
  const { payload } = action

  switch (action.type) {
    case actionTypes.PLAY_VIDEO:
      return {
        ...stopAllVideos(state),
        [payload]: 'playing'
      }
    case actionTypes.STOP_VIDEO:
      return {
        ...state,
        [payload]: 'stopped'
      }
    default:
      return state
  }
}

export default reducer
