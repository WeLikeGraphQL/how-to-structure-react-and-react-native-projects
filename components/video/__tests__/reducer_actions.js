import { actionCreators } from '../actions'
import reducer from '../reducer'

let state = {}

test('1. Play 1st video', () => {
  state = reducer(state, actionCreators.playVideoAction('videoId1'))
  expect(state).toMatchSnapshot()
})

test('2. Play 2nd video', () => {
  state = reducer(state, actionCreators.playVideoAction('videoId2'))
  expect(state).toMatchSnapshot()
})

test('3. Play 3rd video', () => {
  state = reducer(state, actionCreators.playVideoAction('videoId3'))
  expect(state).toMatchSnapshot()
})

test('4. Play 1st video again', () => {
  state = reducer(state, actionCreators.playVideoAction('videoId1'))
  expect(state).toMatchSnapshot()
})

test('5. Stop 1st video', () => {
  state = reducer(state, actionCreators.stopVideoAction('videoId1'))
  expect(state).toMatchSnapshot()
})
