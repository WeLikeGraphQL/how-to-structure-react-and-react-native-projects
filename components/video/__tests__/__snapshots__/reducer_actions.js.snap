// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1. Play 1st video 1`] = `
Object {
  "videoId1": "playing",
}
`;

exports[`2. Play 2nd video 1`] = `
Object {
  "videoId1": "stopped",
  "videoId2": "playing",
}
`;

exports[`3. Play 3rd video 1`] = `
Object {
  "videoId1": "stopped",
  "videoId2": "stopped",
  "videoId3": "playing",
}
`;

exports[`4. Play 1st video again 1`] = `
Object {
  "videoId1": "playing",
  "videoId2": "stopped",
  "videoId3": "stopped",
}
`;

exports[`5. Stop 1st video 1`] = `
Object {
  "videoId1": "stopped",
  "videoId2": "stopped",
  "videoId3": "stopped",
}
`;
